### PR TITLE
add Gatekeeper guide and sample configs

### PR DESCRIPTION
**Pull request by skriss**
Signed-off-by: Steve Kriss <krisss@vmware.com>

replaces #2791 
updates #2742 

To do:
- [x] some more testing of constraints
- [x] test the guide again
- [ ] add more example constraints (to be done in separate PRs)
- [ ] any other spots in docs to add refs to the Gatekeeper config?

### CODE REVIEW

**Comment by skriss** [comment]
Siteproof job failing due to #2032 (links to pages being added in the current PR fail because siteproof checks the external URL).

File: site/_guides/gatekeeper.md
Comment (bgagnon) [inline site/_guides/gatekeeper.md:82]: This is really interesting! I had assumed that Gatekeeper could only see the resource being admitted, plus some contextual data needed for ACLs. This idea of caching resources opens up a whole world of possibilities.

File: site/_guides/gatekeeper.md
Comment (bgagnon) [inline site/_guides/gatekeeper.md:134]: Since Contour is especially good in multi-tenant situations, I'd be interested to see more mentions of namespaces in the Gatekeeper/Contour presentation.

* are constraints namespace-specific by default; can they be made cluster wide
* how constraints at the cluster level interact with namespace-level ones
* what is the order of precedence
* etc.

File: site/_guides/gatekeeper.md
Comment (bgagnon) [inline site/_guides/gatekeeper.md:293]: I think the example should specify `metadata.namespace` field explicitly and a comment below should make the scoping of the constraint extra clear. In other words, if I create this `HTTPProxyRetryCountRange` in namespace `foo`, does it only apply to `HTTPProxy` objects in that namespace? Does it depend on the content of the policy?

File: site/_guides/gatekeeper.md
Comment (skriss) [reply site/_guides/gatekeeper.md:134]: That's great feedback. AFAIK, both `ConstraintTemplates` and `Constraints` must be cluster-scoped, and there's not any other way in Gatekeeper to specify that constraints only apply to specific namespaces. I'll double-check on that to make sure I'm correct but I believe that's the case.

However, you *can* add logic to the rego to effectively make certain constraints namespace-specific -- you'd add a `namespace` parameter to the `ConstraintTemplate`, and then in the rego, you'd add logic to only check `HTTPProxy` objects in that namespace (if specified).

I'll look at getting an example or two of this set up and covering it in the guide. Thanks again for the review!

**Comment by skriss** [comment]
Just pushed a new commit with some edits to demonstrate namespace-scoped Constraints, but I'll re-review and test in the morning.

**Review by youngnick — APPROVED** [review]
LGTM, assuming others are okay.

File: examples/gatekeeper/04-policy-constraint-templates.yml
Comment (jpeach) [inline examples/gatekeeper/04-policy-constraint-templates.yml:106]: I think that this would make the validation more readable:

```
msg := sprintf("%q timeout must have value less than or equal to %q", [input.parameters.field, input.parameters.max])
```

File: examples/gatekeeper/04-policy-constraint-templates.yml
Comment (jpeach) [inline examples/gatekeeper/04-policy-constraint-templates.yml:91]: Suggest you hoist this up to package scope so you doesn't need to be repeated in all the rules.

File: examples/gatekeeper/04-policy-constraint-templates.yml
Comment (jpeach) [inline examples/gatekeeper/04-policy-constraint-templates.yml:59]: hyphenate?

File: examples/gatekeeper/04-policy-constraint-templates.yml
Comment (jpeach) [inline examples/gatekeeper/04-policy-constraint-templates.yml:75]: Is it syntactically possible to apply an enum constraint on this field to enforce getting the correct field name? Could be a separate PR.

File: examples/gatekeeper/04-policy-constraint-templates.yml
Comment (jpeach) [inline examples/gatekeeper/04-policy-constraint-templates.yml:49]: Suggest:
```
msg := sprintf("retry count must be greater than or equal to %q", [input.parameters.min])
```

File: examples/gatekeeper/04-policy-constraint-templates.yml
Comment (jpeach) [inline examples/gatekeeper/04-policy-constraint-templates.yml:46]: Suggest hoisting this out of the rule?

File: site/_guides/gatekeeper.md
Comment (jpeach) [inline site/_guides/gatekeeper.md:439]: I really like the idea of the project hosting a growing collection of templates and constraints. What I would suggest to make this more manageable is to separate each YAML object into a separate file. I think that your classification of validations and policies makes sense, so we can continue it, maybe like this:

```
002-validation-template-Foo-Name.yaml
002-validation-template-Bar-Name.yaml
003-validation-constraint-Foo.yaml
003-validation-constraint-Bar.yaml
004-policy-template-A.yaml
004-policy-template-B.yaml
```

This preserves the classification, but makes it a bit easier to pick and choose.

**Review by jpeach — APPROVED** [review]
Overall, I think that this looks great and should be pretty helpful for people who are interested in admission control polices.

I made a few suggestions, but there's no need to block the PR on them.

File: examples/gatekeeper/01-sync-config.yml
Comment (skriss) [reply examples/gatekeeper/01-sync-config.yml:10]: Unfortunately `Config` is a singleton resource for gatekeeper - see https://github.com/open-policy-agent/gatekeeper/blob/master/pkg/controller/config/config_controller.go#L193-L196 - and the one instance of it has to be named `config`.

File: examples/gatekeeper/02-validation-constraint-templates.yml
Comment (skriss) [reply examples/gatekeeper/02-validation-constraint-templates.yml:14]: It *could*, but Gatekeeper requires that the name of the `ConstraintTemplate` be the lowercase equivalent of the `Kind` of the CRD that it generates. So if we hyphenate the names, then the `Kind` names have to be hyphenated as well, which is very nonstandard for Kubernetes kinds.

File: examples/gatekeeper/01-sync-config.yml
Comment (skriss) [reply examples/gatekeeper/01-sync-config.yml:11]: See https://github.com/projectcontour/contour/pull/2814#discussion_r477420409

File: examples/gatekeeper/02-validation-constraint-templates.yml
Comment (skriss) [reply examples/gatekeeper/02-validation-constraint-templates.yml:16]: yep!

File: examples/gatekeeper/04-policy-constraint-templates.yml
Comment (skriss) [reply examples/gatekeeper/04-policy-constraint-templates.yml:75]: I think it should be possible, I'll look into it

**Comment by skriss** [comment]
Addressed most comments, will run through one more round of testing before being ready to merge.

File: examples/gatekeeper/04-policy-constraint-templates.yml
Comment (skriss) [reply examples/gatekeeper/04-policy-constraint-templates.yml]: Done (had to construct a set since this uses iteration, rather than just a scalar variable)

File: site/_guides/gatekeeper.md
Comment (skriss) [reply site/_guides/gatekeeper.md:446]: I ended up creating two subdirectories, one for `validations` and one for `policies`, and then split things out to a resource per file, per your suggestion.

**Pull request merged**
